File: binary_tree.py
def add(node, val):
    new_node = [ val, [], []]
    if node:
        left, right = node[1:]
        if not left:
            left.extend(new_node)
        elif not right:
            right.extend(new_node)
        else:
            add(left, val)
    else:
        node.extend(new_node)
    return node

def traverse_print(t, order):
    if not t:
        return
    print t[0]
    traverse_print(t[1], order)
    traverse_print(t[2], order)

def traverse_yield(t, order):
    if not t:
        return
    yield t[0]
    for l  in traverse_yield(t[1], order):
        yield l
    for r in traverse_yield(t[2], order):
        yield r

def traverse_global_var(t):
    def helper(t):
        if not t:
            return
        result.append(t[0])
        helper(t[1])
        helper(t[2])
    result = []
    helper(t)
    return result

File: cpp/binary_tree.cpp
#include "binary_tree.h"

#include <boost/lexical_cast.hpp>
#include <iostream>

void add(Node *n, int val) {
    Node *new_node = new Node(val);
    new_node->left = 0;
    new_node->right = 0;
    if (!n->left) {
        n->left = new_node;
        return;
    }
    if (!n->right) {
        n->right = new_node;
        return;
    }
    add(n->left, val);
}

void parse_print(Node *n) {
    std::cout << n->value;
    if (n->left) {
        parse_print(n->left);
    }
    if (n->right) {
        parse_print(n->right);
    }
}


void parse_helper(Node *n, std::string &result) {
    result = result + boost::lexical_cast<std::string>(n->value);
    if (n->left) {
        parse_helper(n->left, result);
    }
    if (n->right) {
        parse_helper(n->right, result);
    }
}

std::string parse(Node *n) {
    std::string result = "";
    parse_helper(n, result);
    return result;
}
